Code/GraphMol/FileParsers/SCSRMolFileParser.cpp

```diff
@@ -49,14 +49,229 @@ class SCSRMol {
+//   errout << "Expected \"TEMPLATE\" at line  " << line;
+//   throw FileParseException(errout.str());
+
+void skipSpaces(const char *&linePtr) {
+  while (*linePtr == ' ') {
+    ++linePtr;  // skip spaces
+  }
+}
+
+std::string getToken(const char *&linePtr, char delim) {
+  skipSpaces(linePtr);
+  unsigned int charCount = 0;
+  while (linePtr[charCount] && linePtr[charCount] != delim) {
+    ++charCount;
+  }
+  std::string res(linePtr, charCount);
+  linePtr += charCount;
+  if (*linePtr == delim) {
+    ++linePtr;  // skip delim
+  }
+  return res;
+}
+
+std::string getQuotedToken(const char *&linePtr) {
+  skipSpaces(linePtr);
+  std::string res;
+  if (*linePtr != '"') {
+    return res;
+  }
+  ++linePtr;  // skip opening quote
+  while (*linePtr && *linePtr != '"') {
+    if (*linePtr == '\\') {
+      // escaped char
+      ++linePtr;
+      if (*linePtr == '\0') {
+        break;
+      }
+    }
+
+    res += *linePtr++;
+  }
+  if (*linePtr != '"') {
+    res = "";  // error: no closing quote
+  } else {
+    ++linePtr;  // skip closing quote
+  }
+
+  return res;
+}
+
+void parseTemplateLine(RWMol *templateMol, std::string lineStr,
+                       unsigned int &line) {
+  PRECONDITION(templateMol, "bad template molecule");
+
+  // TEMPLATE 1 AA/Cya/Cya/ NATREPLACE=AA/A COMMENT=comment
+  // FULLNAME=fullname CATEGORY=cat UNIQUEID=uniqueid CASNUMBER=xxxx,
+  // COLLABORATOR=col, PROTECTION=prot
+
+  // other attributes are allowed.  We capture them and ignore them, except
+  // for writing them back out
+
+  if (lineStr.substr(0, 9) != "TEMPLATE ") {
+    std::ostringstream errout;
+    errout << "Expected \"TEMPLATE\" at line  " << line;
+    throw FileParseException(errout.str());
+  }
+
+  const char *linePtr = lineStr.c_str() + 9;
+  std::string token = getToken(linePtr, ' ');  // Template ID
+  if (token.empty()) {
+    std::ostringstream errout;
+    errout << "Expected a Template ID at line  " << line;
+    throw FileParseException(errout.str());
+  }
+  // get the class and template names
+
+  token = getToken(linePtr, ' ');  // Template ID
+  if (token.empty()) {
+    std::ostringstream errout;
+    errout
+        << "Type/Name(s) string of the form \"AA/Gly/G/\" was not found at line  "
+        << line;
+    throw FileParseException(errout.str());
+  }
+
+  // get the class and template names from the token
+  std::vector<std::string> subTokens;
+  boost::algorithm::split(subTokens, token, boost::algorithm::is_any_of("/"));
+  if (subTokens.size() < 3) {
+    std::ostringstream errout;
+    errout << "Type/Name(s) string is not of the form \"AA/Gly/G/\" at line  "
+           << line;
+    throw FileParseException(errout.str());
+  }
+
+  templateMol->setProp(common_properties::molAtomClass, subTokens[0]);
+
+  std::vector<std::string> templateNames;
+  for (unsigned int i = 1; i < subTokens.size(); ++i) {
+    if (subTokens[i] != "") {
+      templateNames.push_back(subTokens[i]);
+    }
+  }
```

Comment:
I should have noticed this the first time through, sorry.

We're trying (slowly) to replace loops like this with STL algorithms since they can make the intent clearer:

```suggestion
  std::copy_if(subTokens.begin() + 1, subTokens.end(),
               std::back_inserter(templateNames),
               [](const std::string &s) { return !s.empty(); });
```

Reply:
OK made this change.  But I don't think that make the intent any clearer ;)